apiml/src/main/java/org/zowe/apiml/filter/StoreAccessTokenInfoWebFilter.java

```diff
@@ -0,0 +1,87 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.io.buffer.DataBufferUtils;
+import org.springframework.http.server.reactive.ServerHttpRequestDecorator;
+import org.springframework.security.web.server.WebFilterExchange;
+import org.springframework.security.web.server.authentication.ServerAuthenticationFailureHandler;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.web.server.WebFilter;
+import org.springframework.web.server.WebFilterChain;
+import org.zowe.apiml.controller.ReactiveAuthenticationController.AccessTokenRequest;
+import org.zowe.apiml.security.common.error.AccessTokenBodyNotValidException;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.util.stream.Collectors;
+
+import static org.zowe.apiml.security.common.filter.StoreAccessTokenInfoFilter.TOKEN_REQUEST;
+
+/**
+ * A reactive WebFilter that parses and stores access token request details from the request body.
+ * <p>
+ * This filter reads the request body expecting a {@link SuccessfulPersonalAccessTokenHandler.AccessTokenRequest}
+ * JSON payload. It extracts and normalizes the scopes, storing the full object in the exchange attributes
+ * under the key {@code TOKEN_REQUEST} for downstream filters to use.
+ * <p>
+ * If the payload is invalid or missing required fields (like scopes), the provided
+ * {@link ServerAuthenticationFailureHandler} is triggered to handle the error response.
+ * <p>
+ * The original body is reconstructed and passed downstream using {@link ServerHttpRequestDecorator}.
+ */
+@Slf4j
+@RequiredArgsConstructor
+public class StoreAccessTokenInfoWebFilter implements WebFilter {
+
+    private final ServerAuthenticationFailureHandler failureHandler;
+    private final ObjectMapper mapper;
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
+        return DataBufferUtils.join(exchange.getRequest().getBody())
+            .flatMap(dataBuffer -> {
+                byte[] bytes = new byte[dataBuffer.readableByteCount()];
+                dataBuffer.read(bytes);
+                DataBufferUtils.release(dataBuffer);
+
+                try {
+                    var accessTokenRequest = mapper.readValue(bytes, AccessTokenRequest.class);
+
+                    var scopes = accessTokenRequest.getScopes();
+                    if (scopes == null || scopes.isEmpty()) {
+                        return failureHandler.onAuthenticationFailure(
+                            new WebFilterExchange(exchange, chain),
+                            new AccessTokenBodyNotValidException("org.zowe.apiml.security.token.accessTokenBodyMissingScopes")
+                        );
+                    }
+
+                    accessTokenRequest.setScopes(scopes.stream().map(String::toLowerCase).collect(Collectors.toSet()));
+                    exchange.getAttributes().put(TOKEN_REQUEST, accessTokenRequest);
```

Comment:
It is great that we store the scopes, but we do not use them anywhere. It should be part of RouteLocator.

Reply:
I'm not a fan of this filter either but this attribute is used in the /access-token/generate endpoint in ReactiveAuthenticationController

Reply:
Ok, thanks, I see. But does it make sense to use this filter and an extra SecurityWebFilterChain if this is needed just for one endpoint?